Syntax highlighting

mr-karan opened the pull request:
Updated on 08/12/2016: https://asciinema.org/a/82541

mr-karan commented:
I fixed the tests. So only one last thing to worry now is.
Earlier if hello\tworld(\t) is tab, was printed like hello ---> world currently it's only printed as hello >> world (>> depends on tab width which is same as before )
Is that a thing of concern? Or can we adapt to the new style? Here's how it looks

![doesntlookbad](https://cloud.githubusercontent.com/assets/5689132/17247648/4293472a-55b2-11e6-9e06-26a879f22fc2.png)

sils1297 commented:
the new style is fine

mr-karan commented:
@Makman2 @manoranjanp Would you like to review this please ? :)

Makman2 commented:
sorry disabled until wednesday^^

manoranjanp commented:
Going through it

sils1297 commented:
@mr-karan does it keep highlighting the sourcerange correctly? E.g. if you run the TODOBear the todo itself will be highlighted, so the user knows where in that line the issue is.

mr-karan commented:
I couldn't find `TODOBear` but I ran it for `HappinessLintBear` and this is how it looks ->

![screencoala](https://cloud.githubusercontent.com/assets/5689132/17289362/0759900a-57f7-11e6-99a2-e2be0bb85101.png)

mr-karan commented:
Okay, so I ran KeywordBear for `TODO` keyword and here's how it looks @sils1297 

Hope that's how you want it? 

![screenshot from 2016-08-01 16-36-22](https://cloud.githubusercontent.com/assets/5689132/17292225/364873d6-5806-11e6-85fc-71a6d31fe200.png)

sils1297 commented:
Sadly not, the sourcerange in those cases spans over the word TODO because it doesn't affect the whole line but just the TODO! Therefore it should be highlighted, right? (It should be highlighted without your patches.)

sils1297 commented:
@mr-karan marking WIP because of that

mr-karan commented:
@sils1297 Fixed it. So earlier what was happening was I was using a plain text highlighter, but ideally we should fallback to our older highlighting mechanism in such cases as it's better. Still I have kept the background highlighting intact :+1: 
You can take a look here 
![fixedcoalal](https://cloud.githubusercontent.com/assets/5689132/17427714/4cbe260e-5b00-11e6-9551-83ce59ccccd2.png)


@Makman2 @manoranjanp I believe this is ready and you can review this. Would love if we can have this ready for 0.8 release.

mr-karan commented:
WIP fixing the coverage

mr-karan commented:
Hey so I have modified the syntax highlighting to look like this. (Red background for sourcerange rather than red colr text) :+1: 

Fixed the coverage and all tests pass, *finally*. @sils1297 @Makman2 @manoranjanp You can finally review this please.

![newsh](https://cloud.githubusercontent.com/assets/5689132/17455670/4c09ff5c-5bdd-11e6-9b50-38eafdd4477a.png)

sils1297 commented:
neat so syntax highlightning uses foreground colours, result highlightiing background?

sils1297 commented:
(please rebase while you're at it)

mr-karan commented:
yes absolutely 
![yess](https://cloud.githubusercontent.com/assets/5689132/17455699/c3da66c0-5bdd-11e6-9747-884e5f7ab84d.png)

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,39 @@
+    styles = {
+        Token: 'bold bg:#eee #111'
+    }
+
```
missing newline

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,39 @@
+    Sub class of ``Style``.
+
+    This class sets styling properties such as ``background color``,
+    ``text-color`` and ``bold`` on the text.
```
this doccomment seems kind of redundand and unneeded I think^^

Makman2 replied:
same for above

mr-karan replied:
do you want me to change them or remove entirely? Also since you wanted to know what these classes do, they are for highlighting the background and using a different foreground color in the highlighted message or sourcerange.

Makman2 replied:
remove them :)

and alright somehow I though they do more than that :3

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +169,46 @@ def print_lines(console_printer,
+        # Get suitable lexer depending on filename's extension and if not found
+        # fallback to TextLexer().
+        try:
+            lexer = get_lexer_for_filename(os.path.relpath(sourcerange.file))
```
do you need the relative path? can't you just feed in `sourcerange.file` itself?

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +169,46 @@ def print_lines(console_printer,
+        except ClassNotFound:
+            lexer = TextLexer()
+        lexer.add_filter(VisibleWhitespaceFilter(spaces="•", tabs=True,
+                                                 tabsize=4))
```
you should use `SpacingHelper.DEFAULT_TAB_WIDTH`

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +169,46 @@ def print_lines(console_printer,
-                tab_dict, FILE_LINES_COLOR)
+            console_printer.print(highlight(
+                 line[:sourcerange.start.column-1],
+                lexer, TerminalTrueColorFormatter()).rstrip("\n"), end='')
```
indentation - add a leading space :)

Makman2 replied:
what is the `rstrip("\n")` for?

mr-karan replied:
`highlight` automatically adds a `("\n")`, so this is the only way I can remove it.

sils1297 replied:
just do `[:-1]` then instead with a comment?

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +169,46 @@ def print_lines(console_printer,
+            console_printer.print(highlight(
+                line[printed_chars:], lexer,
+                TerminalTrueColorFormatter(style=BackgroundSourceRange)).rstrip(
+                                                                  "\n"), end='')
```
is the indentation correct? :O this is kind of a corner case...

mr-karan replied:
`coala -S autoapply=True` does this automatically even if I manually change

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -226,8 +242,11 @@ def print_result(console_printer,
+    lexer = TextLexer()
+    result.message = (highlight(result.message, lexer,
+                                TerminalTrueColorFormatter(
+                                                      style=BackgroundMessage)))
```
indentation:
4 spaces from beginning of `TerminalTrueColorFormatter`

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -226,8 +242,11 @@ def print_result(console_printer,
+                                TerminalTrueColorFormatter(
+                                                      style=BackgroundMessage)))
     console_printer.print(format_lines(result.message), delimiter="\n")
-
```
unnecessary codestyle change + I would keep it for readability :3

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -37,6 +45,30 @@
+    '''
+    styles = {
+        Token: 'bold bg:#eee #111'
+    }
```
these are copied, please import them from `ConsoleInteraction`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -162,6 +195,7 @@ def hijack_print(text, *args, **kwargs):
                                      dict(sh.yield_tab_lengths(test_string)),
                                      "red")
         self.assertEqual(printer.string, ".he>llo..world.")
+    '''
```
please remove the test case :3

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -439,8 +473,10 @@ def test_print_result_no_input(self):
-|    | message
-""")
+|    | {}
+""".rstrip("\r\n").format(highlight("message", TextLexer(),
```
why `rstrip()`? if you want to remove returns, just put `"""` onto the previous line :+1:

mr-karan replied:
The output doesn't match, because highlight seems to be adding a "\n" and then return a `str` object. So I thought the best way is to use `rstrip()` to get rid of it.

Makman2 replied:
but you are first stripping the test string, and they you highlight it.

mr-karan replied:
Oh stupid me. Thanks, I ignored that and was thinking you're talking about the other rstrip. Fixed now

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,11 +504,17 @@ def test_print_results_project_wide(self):
+                                                      TerminalTrueColorFormatter
+                                                      (style=BackgroundMessage))
+                          ),
+                                                      stdout.getvalue())
```
alignment: 4 spaces indented starting from `self.assertEqual`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,11 +504,17 @@ def test_print_results_project_wide(self):
+                "".format(STR_PROJECT_WIDE, highlight("message", TextLexer(),
+                                                      TerminalTrueColorFormatter
+                                                      (style=BackgroundMessage))
+                          ),
```
what about reformatting it like this:

```python
self.assertEqual(
    "".format(
        STR_PROJECT_WIDE,
        highlight("message", TextLexer(),
                  TerminalTrueColorFormatter(style=BackgroundMessage))),
    stdout.getvalue())
```

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,13 +555,22 @@ def test_print_results_for_file(self):
+        self.maxDiff = None
+        lexer = TextLexer()
+        lexer.add_filter(VisibleWhitespaceFilter(spaces="•", tabs=True,
+                                                 tabsize=4))
```
wanna setup `lexer` inside `setUp()`?

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -485,10 +527,16 @@ def test_print_results_for_file(self):
+                          TextLexer(),
+                          TerminalTrueColorFormatter(
+                                                         style=BackgroundMessage
+                                                             )).rstrip("\r\n")),
```
indentation (see above)

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -535,18 +592,30 @@ def test_print_results_sorting(self):
+        lexer = TextLexer()
+        lexer.add_filter(VisibleWhitespaceFilter(spaces="•", tabs=True,
+                                                 tabsize=4))
+        self.maxDiff = None
```
disable that please again :)

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -560,24 +629,40 @@ def test_print_results_multiple_ranges(self):
+                    style=BackgroundSourceRange)).rstrip("\n"),
+                highlight('line 7', lexer, TerminalTrueColorFormatter(
+                    style=BackgroundSourceRange)).rstrip("\n"),
+                highlight("Clone Found", TextLexer(),
```
maybe make a method inside the test class out of this `highlight` usage?

\+ why suddenly a new `TextLexer` instance?

mr-karan replied:
`TextLexer` is for result messages. `lexer` is for `sourcerange`. The highlighter is different and also `TextLexer` doesn't need bullet marks for `spaces/tabs`. Hence the 2 different lexers.

Makman2 commented:
ack dd26c25

mr-karan commented:
@Makman2 The next iteration is ready :+1:

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,28 @@
 from coalib.settings.Setting import Setting
 
+from pygments import highlight
+from pygments.util import ClassNotFound
```
this import needs to be the last one (alphabetical order :D)

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,47 @@ def print_lines(console_printer,
+            lexer = TextLexer()
+        s = SpacingHelper.DEFAULT_TAB_WIDTH
+        lexer.add_filter(VisibleWhitespaceFilter(spaces="•", tabs=True,
+                                                 tabsize=s))
```
please inline `s` :P

mr-karan replied:
I did it because unable follow the 80 chars rule otherwise. Also it was done [here](https://github.com/coala-analyzer/coala/blob/master/coalib/output/ConsoleInteraction.py#L173) so I thought I can use this style?

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,47 @@ def print_lines(console_printer,
-                tab_dict, FILE_LINES_COLOR)
+            console_printer.print(highlight(
+                 line[:sourcerange.start.column-1],
+                 lexer, TerminalTrueColorFormatter()).rstrip("\n"), end='')
```
ah both are overindented by one space :3
sorry last time it looked like this line was this problem, but it was the previous one :3

Makman2 replied:
ah both are overindented by one space :3
sorry last time it looked like this line was this problem, but it was the previous one :3

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,47 @@ def print_lines(console_printer,
+            console_printer.print(highlight(
+                line[printed_chars:sourcerange.end.column-1],
+                lexer, TerminalTrueColorFormatter(style=BackgroundSourceRange))
+                .rstrip("\n"), end='')
```
what about using `[:-1]` like suggested?

mr-karan replied:
Do you mean `line[:-1]` ? It's not correct, will show repeated characters. But I'm not sure what do you actually mean ?

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -16,10 +16,12 @@
-    print_spaces_tabs_in_unicode, show_bear, show_bears,
+    show_bear, show_bears,
     ask_for_action_and_apply, print_diffs_info,
     show_language_bears_capabilities)
```
freed space, you could rewrap imports :)

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -29,6 +31,14 @@
 
+from pygments import highlight
+from pygments.formatters import TerminalFormatter
+from pygments.formatters import TerminalTrueColorFormatter
```
-->
```
from pygments.formatters import (TerminalFormatter
                                 TerminalTrueColorFormatter)
```

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,28 @@
+from pygments import highlight
+from pygments.util import ClassNotFound
+from pygments.formatters import TerminalFormatter
+from pygments.formatters import TerminalTrueColorFormatter
```
-->
```
from pygments.formatters import (TerminalFormatter,
                                 TerminalTrueColorFormatter)
```

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -127,42 +137,16 @@ def setUp(self):
+        s = SpacingHelper.DEFAULT_TAB_WIDTH
+        self.lexer.add_filter(VisibleWhitespaceFilter(spaces="•",
+                                                      tabs=True,
+                                                      tabsize=s))
```
I would inline `s` here too :+1:

mr-karan replied:
reason to not mentioned above^

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -439,8 +423,9 @@ def test_print_result_no_input(self):
-""")
+|    | {}""".format(highlight("message", TextLexer(),
+                                     TerminalTrueColorFormatter(
+                                    style=BackgroundMessage))))
```
this is weird indentation :3

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,9 +453,13 @@ def test_print_results_project_wide(self):
-                "\n".format(STR_PROJECT_WIDE),
-                stdout.getvalue())
+                "\n{0}\n|    | [NORMAL] origin:\n|    | {1}"
+                     "".format(
```
what about
```
"\n{0}\ņ.......".format(
    arg1,
    arg2,
    ...
```

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,9 +453,13 @@ def test_print_results_project_wide(self):
+                         highlight("message", TextLexer(),
+                                   TerminalTrueColorFormatter(
+                                        style=BackgroundMessage))),
+                        stdout.getvalue())
```
wrong indentation, needs to be 4 spaces indented relative to `self.assertEqual`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,11 +501,15 @@ def test_print_results_for_file(self):
+                                style=BackgroundSourceRange)).rstrip("\r\n"),
+                highlight("Trailing whitespace found", TextLexer(),
+                          TerminalTrueColorFormatter(style=BackgroundMessage
+                                                     )).rstrip("\r\n")),
```
you should really make some shortcut functions for the usages of `highlight`^^

mr-karan commented:
@Makman2 I have made the changes and felt where the change can't be done, I have mentioned. Take a look, please. 
Thanks :+1:

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -29,6 +30,14 @@
+from pygments.lexers import TextLexer
+from pygments.style import Style
+from pygments.token import Token
+
```
missing newline

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -439,8 +422,10 @@ def test_print_result_no_input(self):
+|    | {}""".format(highlight("message", TextLexer(),
+                                     TerminalTrueColorFormatter(
+                                                        style=BackgroundMessage)
+                              )))
```
-->
```python
format(hightligh("message", TextLexer(),
                 TerminalTrueColorFormatter(
                     style=BackgroundMessage)))
```

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,28 @@
+    }
+
+
+class BackgroundMessage(Style):
```
not sure, maybe add `Style` to the name of these classes? As when you use them the name doesn't really look like a style :3

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,11 +453,17 @@ def test_print_results_project_wide(self):
-                "\n{}\n|    | [NORMAL] origin:\n|    | message"
-                "\n".format(STR_PROJECT_WIDE),
-                stdout.getvalue())
+                    "\n{0}\n|    | [NORMAL] origin:\n|    | {1}""".format(
```
two redundant quotes right before format

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,11 +453,17 @@ def test_print_results_project_wide(self):
+                             highlight("message", TextLexer(),
+                                       TerminalTrueColorFormatter(
+                                            style=BackgroundMessage))),
+                            stdout.getvalue())
```
-->
```python
self.assertEqual(
    "\n{}\n|    | [NORMAL] origin:\n|    | {}".format(
        STR_PROJECT_WIDE,
        highlight("message", TextLexer(),
                  TerminalTrueColorFormatter(style=BackgroundMessage))),
    stdout.getvalue())
```

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -485,11 +476,14 @@ def test_print_results_for_file(self):
-|    | Trailing whitespace found
-""",
-                             stdout.getvalue())
+|    | {1}
```
no need to explicitly number the placeholders, as you use them only once. `{0}` or `{1}` --> `{}`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,11 +453,17 @@ def test_print_results_project_wide(self):
     def test_print_results_for_file(self):
+        def highlightoutput(text, lexer=self.lexer):
+            return highlight(text, lexer, TerminalTrueColorFormatter(
+                style=BackgroundSourceRange)).rstrip("\n")
```
you should also introduce a `style` parameter with a default, then you can use this function on even more places :+1:

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -485,11 +476,14 @@ def test_print_results_for_file(self):
+""".format(highlightoutput('line 2'),
+                highlight("Trailing whitespace found", TextLexer(),
+                          TerminalTrueColorFormatter(style=BackgroundMessage)
+                          ).rstrip("\r\n")),
```
three lines above overindented by one space

Makman2 replied:
ah no by 5 spaces :3

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,13 +501,20 @@ def test_print_results_for_file(self):
-|    | Trailing whitespace found
-""",
-                             stdout.getvalue())
+|    | {1}
```
here too: `{0}` and `{1}` --> `{}`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,13 +501,20 @@ def test_print_results_for_file(self):
+""".format(highlightoutput('line 5'),
+                highlight("Trailing whitespace found", TextLexer(),
+                          TerminalTrueColorFormatter(style=BackgroundMessage
+                                                     )).rstrip("\r\n")),
```
also overindented by 5 spaces

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -535,18 +536,25 @@ def test_print_results_sorting(self):
+""".format(highlightoutput('\t'),
+                highlight("Trailing whitespace found", TextLexer(),
+                          TerminalTrueColorFormatter(style=BackgroundMessage)
+                          ).rstrip("\r\n"),
```
overindented by 5 spaces

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,13 +501,20 @@ def test_print_results_for_file(self):
     def test_print_results_sorting(self):
+        def highlightoutput(text):
+            return highlight(text, self.lexer, TerminalTrueColorFormatter(
+                style=BackgroundSourceRange)).rstrip("\n")
```
move that function to class level, then you can remove all other declarations :)

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -560,24 +568,36 @@ def test_print_results_multiple_ranges(self):
+                highlightoutput('line 7'),
+                highlight("Clone Found", TextLexer(),
+                          TerminalTrueColorFormatter(
+                                            style=BackgroundMessage)).rstrip(
```
indentation should be 4 spaces relative to TerminalTrueColorFormatter

mr-karan commented:
Review please :3 Did the changes. @Makman2

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
+            lexer = TextLexer()
+        lexer.add_filter(VisibleWhitespaceFilter(
+                            spaces="•", tabs=True,
+                            tabsize=SpacingHelper.DEFAULT_TAB_WIDTH))
```
indentation: \+ 1 space

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
+            console_printer.print(highlight(
+                line[printed_chars:sourcerange.end.column-1],
+                lexer, TerminalTrueColorFormatter(
+                        style=BackgroundSourceRangeStyle))[:-1], end='')
```
indentation: 4 spaces starting from `lexer`

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -226,6 +233,10 @@ def print_result(console_printer,
+    lexer = TextLexer()
+    result.message = (highlight(result.message, lexer,
+                                TerminalTrueColorFormatter(
+                                    style=BackgroundMessageStyle)))
```
redundant braces
\+
what about `[:-1]`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -439,8 +428,9 @@ def test_print_result_no_input(self):
 |    | [NORMAL] origin:
-|    | message
-""")
+|    | {}""".format(ConsoleInteractionTest.highlightoutput(
```
if you want you can use `self` instead of `ConsoleInteractionTest` to shorten the stmt^^

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -127,42 +137,21 @@ def setUp(self):
+    def highlightoutput(text, lexer=TextLexer(),
+                        style=BackgroundSourceRangeStyle):
+        return highlight(text, lexer, TerminalTrueColorFormatter(
+            style=style))
```
you forgot to make it a `@staticmethod`

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -468,8 +458,11 @@ def test_print_results_project_wide(self):
+                    STR_PROJECT_WIDE,
+                    ConsoleInteractionTest.highlightoutput(
+                                            "message",
+                                            style=BackgroundMessageStyle)),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,11 +503,14 @@ def test_print_results_for_file(self):
+""".format(ConsoleInteractionTest.highlightoutput('line 5', self.lexer)[:-1],
+                ConsoleInteractionTest.highlightoutput(
+                                        "Trailing whitespace found",
+                                        style=BackgroundMessageStyle))[:-1],
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -593,12 +610,16 @@ def test_print_results_missing_file(self):
+                             "|    | {0}".format(
+                                        ConsoleInteractionTest.highlightoutput(
+                                            "msg", style=BackgroundMessageStyle
+                                            )),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -612,14 +633,20 @@ def test_print_results_missing_line(self):
+                                ConsoleInteractionTest.highlightoutput(
+                                                "msg",
+                                                style=BackgroundMessageStyle),
+                                STR_LINE_DOESNT_EXIST),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -634,7 +661,9 @@ def test_print_results_without_line(self):
-                "|    | msg\n",
+                "|    | {}".format(ConsoleInteractionTest.highlightoutput(
+                                    "msg",
+                                    style=BackgroundMessageStyle)),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -127,42 +137,21 @@ def setUp(self):
+        self.lexer.add_filter(VisibleWhitespaceFilter(
+                              spaces="•",
+                              tabs=True,
+                              tabsize=SpacingHelper.DEFAULT_TAB_WIDTH))
```
indentation

sils1297 commented:
@mr-karan your commits should be squashed, the requirements are there for the other functionality, they don't make sense on it's own

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
     for i in range(sourcerange.start.line, sourcerange.end.line + 1):
         console_printer.print(format_lines(lines='', line_nr=i),
                               color=FILE_LINES_COLOR,
                               end='')
```
what is this weird statement for anyway? :/

mr-karan replied:
hey this is for printing the line number

![there](https://cloud.githubusercontent.com/assets/5689132/17547762/74bd9b12-5f06-11e6-8337-428360f6fa4f.png)

And if you remove that line:

![missing](https://cloud.githubusercontent.com/assets/5689132/17547780/7f4c65cc-5f06-11e6-837c-0cf88ba0bf41.png)

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
+            console_printer.print(highlight(
+                line[printed_chars:], lexer,
+                TerminalTrueColorFormatter(
+                    style=BackgroundSourceRangeStyle))[:-1], end='')
```
hm does the highlighting work nicely? After all the lexer gets only small potions of the code at the time and for good highlighting you usually need context as well don't you?

mr-karan replied:
Hm so what I feel is that it checks for that language's special reserved keywords like `while` `in` `for` `range` `if` `else` and stuff like that. Also distinguish on data types etc. That works well even if it's one line at a time. If there are other cases, then that would be a limitation of Pygments as well, I believe. But I feel syntax highlighting presently works well :+1: 

I did a small experiment in python shell by putting the string at once, to see if there's a difference but the result is pretty same and (expected) too.
in coala
![screenshot from 2016-08-10 14-47-32](https://cloud.githubusercontent.com/assets/5689132/17548429/82995700-5f09-11e6-88fe-a0e0818f78b6.png)

in python shell
 - Only one line supplied.
![screenshot from 2016-08-10 14-53-43](https://cloud.githubusercontent.com/assets/5689132/17548609/41b11c36-5f0a-11e6-8075-bd22db2a2f63.png)

 - Full file supplied
![screenshot from 2016-08-10 14-52-33](https://cloud.githubusercontent.com/assets/5689132/17548576/17bc995a-5f0a-11e6-881e-79d631bbf2da.png)

mr-karan commented:
@Makman2 Think it's ready for a final iteration? I did also add a comment to explain a part in an own commit which @sils1297 [mentioned](https://gitter.im/coala-analyzer/coala?at=57aafe9aff82b9fc7e718e72) ^

sils1297 commented:
ack https://github.com/coala-analyzer/coala/pull/2569/commits/5c9f27dc2abbc90be4b182159d3c02cff01e3483

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
-        s = SpacingHelper(tab_width)
-        tab_dict = dict(s.yield_tab_lengths(line))
+        # Get suitable lexer depending on filename's extension and if not found
+        # fallback to TextLexer().
```
IMO this comment is useless and its self explanatory

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -164,35 +158,48 @@ def print_lines(console_printer,
+                             spaces="•", tabs=True,
+                             tabsize=SpacingHelper.DEFAULT_TAB_WIDTH))
+        # highlight() combines lexer and formatter to output a ``str``
+        # object.
```
how about `term_highlight = lambda string: highlight(string, TerminalTrueColorFormatter())[:-1]` to reduce the duplication?

mr-karan commented:
reack https://github.com/coala-analyzer/coala/pull/2569/commits/1957dd48a0155868f586efe33f7721f8580d60e9

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -165,35 +159,43 @@ def print_lines(console_printer,
                             lines to print.
     """
     for i in range(sourcerange.start.line, sourcerange.end.line + 1):
-        # Print affected file's line number in the sidebar.
```
why are you remoing that line?!

sils1297 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -227,7 +229,11 @@ def print_result(console_printer,
+    lexer = TextLexer()
+    result.message = highlight(result.message, lexer,
+                               TerminalTrueColorFormatter(
+                                    style=BackgroundMessageStyle))[:-1]
```
add a third lexer parameter (optional) to your lambda so you can use it here as well?

mr-karan commented:
@Makman2 Hey, fixed up my stuff. Please take a look now :)

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,36 @@
+        return highlight(text, lexer, TerminalTrueColorFormatter(
+                             style=style))[:-1]
+    else:
+        return highlight(text, lexer, TerminalTrueColorFormatter())[:-1]
```
what happens if you pass `None` as style?

btw -->
```python
formatter = TerminalTrueColorFormatter()
if style:
    highlighted = highligh(text, lexer, formatter, style=style)
else:
    highlighted = highlight(text, lexer, formatter)
return highlighted[:-1]
```

Makman2 replied:
ah no wait:
```python
if style:
    formatter = TerminalTrueColorFormatter(style=style)
else:
    formatter = TerminalTrueColorFormatter()
return highlight(text, lexer, formatter)[:-1]
```

mr-karan replied:
If i pass None then it's an error:
```
>>> highlight("hi",PythonLexer(),TerminalTrueColorFormatter(style=None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/karan/Work/gsoc/coding-period/venv/lib/python3.5/site-packages/pygments/formatters/terminal256.py", line 115, in __init__
    self._setup_styles()  # convert selected style's colors to term. colors
  File "/home/karan/Work/gsoc/coding-period/venv/lib/python3.5/site-packages/pygments/formatters/terminal256.py", line 270, in _setup_styles
    for ttype, ndef in self.style:
TypeError: 'NoneType' object is not iterable
```

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -23,6 +23,36 @@
+    }
+
+
+def highlightoutput(text, lexer=TextLexer(), style=None):
```
maybe just `highlight_text`, as your first param is `text`? :)

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -169,31 +171,36 @@ def print_lines(console_printer,
-                console_printer, line[printed_chars:], tab_dict,
-                HIGHLIGHTED_CODE_COLOR, printed_chars)
+            console_printer.print(highlightoutput(
+                    line[printed_chars:], lexer), end='')
```
indentation: though you open another scope, it's indented only by 4 spaces :+1:

Makman2 commented on coalib/output/ConsoleInteraction.py:
```diff
@@ -227,7 +234,10 @@ def print_result(console_printer,
-    console_printer.print(format_lines(result.message), delimiter="\n")
+    lexer = TextLexer()
+    result.message = highlightoutput(
+                         result.message, lexer, BackgroundMessageStyle)
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -507,11 +495,13 @@ def test_print_results_for_file(self):
+|    | {}
+""".format(highlightoutput('line 5', self.lexer),
+                highlightoutput("Trailing whitespace found",
+                                style=BackgroundMessageStyle)),
```
indentation:
```
format(A,
       B)
```

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -535,16 +525,19 @@ def test_print_results_sorting(self):
+""".format(highlightoutput('\t', self.lexer),
+                highlightoutput("Trailing whitespace found",
+                                style=BackgroundMessageStyle),
+                highlightoutput('line 5\t', self.lexer)),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -560,25 +553,32 @@ def test_print_results_multiple_ranges(self):
+                highlightoutput('line 6', self.lexer),
+                highlightoutput('line 7', self.lexer),
+                highlightoutput("Clone Found",
+                                style=BackgroundMessageStyle)),
```
indentation

Makman2 commented on tests/output/ConsoleInteractionTest.py:
```diff
@@ -593,12 +593,16 @@ def test_print_results_missing_file(self):
+                             "|    | {0}\n"
+                             .format(highlightoutput(
+                                         "msg",
+                                         style=BackgroundMessageStyle)),
```
indentation

sils1297 commented:
ack https://github.com/coala-analyzer/coala/pull/2569/commits/14ece91722cdcdb68303fd72996bdcaf65224bfc

sils1297 commented:
ack https://github.com/coala-analyzer/coala/pull/2569/commits/4e20e9ae1f601f1c8524839f2c01b602ca8967cb

sils1297 commented:
@rultor merge

rultor commented:
> @rultor merge

@sils1297 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/9268-239417714)

rultor commented:
> @rultor merge

@sils1297 Done! FYI, the full log is [here](http://www.rultor.com/t/9268-239417714) (took me 1min)
